File: Courses/PHP Web Basics - май 2019/16.ExerciseEncapsulationInheritance/01.03.ShoppingSpree/Product.php
<?php


class Product
{
    /**
     * @var string
     */
    private $name;
    /**
     * @var float
     */
    private $cost;

    /**
     * Product constructor.
     * @param string $name
     * @param float $cost
     * @throws Exception
     */
    public function __construct(string $name, float $cost)
    {
        $this->setName($name);
        $this->setCost($cost);
    }

    /**
     * @return string
     */
    public function getName(): string
    {
        return $this->name;
    }

    /**
     * @param string $name
     * @throws Exception
     */
    private function setName(string $name): void
    {
        if($name == ''){
            throw new Exception('Name cannot be empty'.PHP_EOL);
        }
        $this->name = $name;
    }

    /**
     * @return float
     */
    public function getCost(): float
    {
        return $this->cost;
    }

    /**
     * @param float $cost
     * @throws Exception
     */
    private function setCost(float $cost): void
    {
        if($cost < 0){
            throw new Exception('Cost cannot be negative'.PHP_EOL);
        }
        $this->cost = $cost;
    }
}

File: Courses/C# OOP/08. Encapsulation Exercise/08. Encapsulation Exercise/04.ShoppingSpree/Product.cs
﻿using System;

namespace ShoppingSpree
{
    public class Product
    {
        private string name;
        private decimal cost;

        public Product(string name, decimal cost)
        {
            Name = name;
            Cost = cost;
        }

        public string Name 
        { 
            get => name;

            set
            {
                if (string.IsNullOrEmpty(value))
                {
                    throw new ArgumentException("Name cannot be empty");
                }
                name = value;
            }
        }
        public decimal Cost 
        { 
            get => cost;
            set
            {
                if (value <= 0)
                {
                    throw new ArgumentException("Money cannot be negative");
                }
                cost = value;
            }
        }

        public override string ToString()
        {
            return Name;
        }
    }
}
